### PR TITLE
Handle infinite/NaN rects in Hero flights. Less exclamation marks.

### LINKED ISSUE #61929: Kept alive Hero throws when the transition begins
`localToGlobal` reports `(NaN, NaN)` when the render object is in the keep alive bucket

```dart
  testWidgets('kept alive Hero does not throw when the transition begins', (WidgetTester tester) async {
    final FocusNode focusNode = FocusNode();
    final GlobalKey<NavigatorState> navigatorKey = GlobalKey<NavigatorState>();

    await tester.pumpWidget(
      MaterialApp(
        navigatorKey: navigatorKey,
        home: Scaffold(
          body: ListView(
            children: <Widget>[
              Hero(
                tag: 'a',
                child: Material(
                  child: MediaQuery(
                    data: const MediaQueryData(),
                    child: TextField(focusNode: focusNode)
                  ),
                ),
              ),
              Container(height: 1000.0),
            ],
          ),
        ),
      ),
    );

    FocusScope.of(tester.element(find.byType(TextField))).requestFocus(focusNode);
    await tester.pump();

    await tester.drag(find.byType(ListView), const Offset(0.0, -1000.0));
    await tester.pump();

    navigatorKey.currentState.push(
      MaterialPageRoute<void>(
        builder: (BuildContext context) {
          return Scaffold(
            body: Center(
              child: Hero(
                tag: 'a',
                child: Material(
                  child: MediaQuery(
                    data: const MediaQueryData(),
                    child: TextField(focusNode: focusNode)
                  ),
                ),
              ),
            ),
          );
        },
      ),
    );
    await tester.pumpAndSettle();
  });
```

### PR DESCRIPTION
**Pull request by LongCatIsLooong**
## Description

The framework throws an exception when an unpaintable Rect is returned by `transformRect` and the like. and the `_buildOverlay` method tries to layout/paint the Rect. 

Removes some exclamation marks.

## Related Issues

Fixes https://github.com/flutter/flutter/issues/61929

## Tests

I added the following tests:

- diverting to a keepalive but unpaintable hero
- kept alive Hero does not throw when the transition begins

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [ ] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/widgets/heroes.dart [hunk 1]
```diff
@@ -849,63 +875,89 @@ class HeroController extends NavigatorObserver {
...
+      return;
+    }
+    assert(navigatorRenderObject.hasSize);
 
-        if (isDiverted)
-          _flights[tag]!.divert(manifest);
-        else
+    // At this point the toHeroes may have been built and laid out for the first time.
+    //
+    // If `fromSubtreeContext` is null, call endFlight on all toHeroes, for good measure.
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/widgets/heroes.dart:901]: Is it possible for `fromSubtreeContext` or `toSubtreeContext` to be null? I'll add tests to cover that if they can be null.

File: packages/flutter/lib/src/widgets/heroes.dart [hunk 2]
```diff
@@ -642,39 +672,36 @@ class _HeroFlight {
...
         newManifest.toHero.startFlight();
-        heroRectTween = _doCreateRectTween(
-            heroRectTween.end,
-            _boundingBoxFor(newManifest.toHero.context, newManifest.toRoute.subtreeContext),
-        );
+        heroRectTween = manifest.createHeroRectTween(begin: heroRectTween.end, end: newManifest.toHeroLocation);
       } else {
         // TODO(hansmuller): Use ReverseTween here per github.com/flutter/flutter/pull/12203.
-        heroRectTween = _doCreateRectTween(heroRectTween.end, heroRectTween.begin);
+        heroRectTween = manifest.createHeroRectTween(begin: heroRectTween.end, end: heroRectTween.begin);
```
[the remark this answers is not in the document]
Comment (chunhtai) [reply packages/flutter/lib/src/widgets/heroes.dart:694]: Does this matter? This will only affect when the divert happens when toHero is right at the hero rect tween end. If that is the case, changing the new manifest may result in a sudden jump.

File: packages/flutter/lib/src/widgets/heroes.dart [hunk 3]
```diff
@@ -506,17 +536,16 @@ class _HeroFlight {
         } else if (toHeroBox.hasSize) {
           // The toHero has been laid out. If it's no longer where the hero animation is
           // supposed to end up then recreate the heroRect tween.
-          final RenderBox? finalRouteBox = manifest!.toRoute.subtreeContext?.findRenderObject() as RenderBox?;
+          final RenderBox? finalRouteBox = manifest.toRoute.subtreeContext?.findRenderObject() as RenderBox?;
           final Offset toHeroOrigin = toHeroBox.localToGlobal(Offset.zero, ancestor: finalRouteBox);
-          if (toHeroOrigin != heroRectTween.end!.topLeft) {
+          if (toHeroOrigin != heroRectTween.end!.topLeft && toHeroOrigin.isFinite) {
```
[the remark this answers is not in the document]
Comment (chunhtai) [reply packages/flutter/lib/src/widgets/heroes.dart:542]: the flight will start only if the manifest is valid and thus toHeroOrigin.isFinite must be true?

This should be an assert?

File: packages/flutter/lib/src/widgets/heroes.dart (hunk 1 above)
Comment (chunhtai) [reply packages/flutter/lib/src/widgets/heroes.dart:902]: I can't think of a case this can happen, but i may be wrong

File: packages/flutter/lib/src/widgets/heroes.dart [hunk 4]
```diff
@@ -443,10 +442,47 @@ class _HeroFlightManifest {
...
+    final CreateRectTween? createRectTween = toHero.widget.createRectTween ?? this.createRectTween;
+    return createRectTween?.call(begin, end) ?? RectTween(begin: begin, end: end);
+  }
+
+  // The bounding box for `context`'s render object,  in `ancestorContext`'s
+  // render object's coordinate space.
+  static Rect _boundingBoxFor(BuildContext context, BuildContext? ancestorContext) {
+    assert(ancestorContext != null);
+    final RenderBox box = context.findRenderObject()! as RenderBox;
+    assert(box != null && box.hasSize && box.size.isFinite);
```
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/heroes.dart:455]: if the fromHeroLocation.size is not finite, this assertion will throw before isValid returns false?

Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/heroes.dart:455]: The PR for the most part is trying to address infinity/NaN introduced by applying a zero matrix. If RenderBox.size contains either infinity/NaN then it's probably an error.

File: packages/flutter/lib/src/widgets/heroes.dart (hunk 3 above)
[the remark this answers is not in the document]
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/heroes.dart]: Changed the implementation so that if that happens the flight gets aborted.

File: packages/flutter/test/widgets/heroes_test.dart [hunk 5]
```diff
@@ -2780,4 +2780,213 @@ Future<void> main() async {
...
+
+
+    expect(findRenderOpacity()?.opacity, anyOf(isNull, 1.0));
+
+    // Scroll to make the Hero invisible.
+    controller.jumpTo(1000);
+    await tester.pump();
+    // The update happens in a build block so we need to keep pumping the
+    // animation to trigger updates.
+    await tester.pump(const Duration(milliseconds: 10));
```
Comment (LongCatIsLooong) [inline packages/flutter/test/widgets/heroes_test.dart:2903]: This is a bit weird. We probably should just let each _HeroFlight listen to the animation controller and update the states in the listener.

File: packages/flutter/lib/src/widgets/heroes.dart (hunk 4 above)
Comment (chunhtai) [reply packages/flutter/lib/src/widgets/heroes.dart:455]: this maybe an extreme case, I have encountered a use case where there are renderobject with zero size. One customer introduce a widget that has various size depends on data load. not that I think we should definitely handle this case, but would like to bring that up for your attention.

File: packages/flutter/lib/src/widgets/heroes.dart [hunk 6]
```diff
@@ -849,63 +882,89 @@ class HeroController extends NavigatorObserver {
...
-        if (isDiverted)
-          _flights[tag]!.divert(manifest);
-        else
+    if (navigatorRenderObject is! RenderBox) {
+      assert(false, 'Navigator $navigator has an invalid RenderObject type ${navigatorRenderObject.runtimeType}.');
+      return;
+    }
+    assert(navigatorRenderObject.hasSize);
+
+    // At this point the toHeroes may have been built and laid out for the first time.
```
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/heroes.dart:906]:
```suggestion
    // At this point, the toHeroes may have been built and laid out for the first time.
```

**Review by chunhtai — COMMENTED** [review]
mostly nits, otherwise looks good

File: packages/flutter/lib/src/widgets/heroes.dart (hunk 4 above)
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/heroes.dart:455]: Sorry I fail to understand the concern here, are you saying if the bounding box is zero-sized it's going to cause problems?

Comment (chunhtai) [reply packages/flutter/lib/src/widgets/heroes.dart:455]: If the renderobject is zero size I think this will throw before the hero flight manifest isvalid returns false

File: packages/flutter/lib/src/widgets/heroes.dart [hunk 7]
```diff
@@ -613,6 +591,33 @@ class _HeroFlight {
...
+      final RenderBox? finalRouteBox = manifest.toRoute.subtreeContext?.findRenderObject() as RenderBox?;
+      final Offset toHeroOrigin = toHeroBox.localToGlobal(Offset.zero, ancestor: finalRouteBox);
+
+      _aborted = !toHeroOrigin.isFinite;
+      if (!_aborted && toHeroOrigin != heroRectTween.end!.topLeft) {
+        final Rect heroRectEnd = toHeroOrigin & heroRectTween.end!.size;
+        heroRectTween = manifest.createHeroRectTween(begin: heroRectTween.begin, end: heroRectEnd);
+      }
+    } else {
+      _aborted = true;
```
[the remark this answers is not in the document]
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/heroes.dart:609]: The reason I left the entire expression in the if clause is that `toHeroBox` does not get type prompted to non-null if I just do `if (!aborted)`

[the remark this answers is not in the document]
Comment (chunhtai) [reply packages/flutter/lib/src/widgets/heroes.dart:609]: I think one additional ! in the if case is fine.

File: packages/flutter/lib/src/widgets/heroes.dart (hunk 4 above)
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/heroes.dart:455]: How so? `Size.zero` seems valid to me, isn't `Size.zero.isFinite` true?

Comment (chunhtai) [reply packages/flutter/lib/src/widgets/heroes.dart:455]: for some reason i keep thinking isfinite = size >0 <infinite, yeah i think this is good

File: packages/flutter/lib/src/widgets/heroes.dart (hunk 7 above)
[the remark this answers is not in the document]
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/heroes.dart:609]: Oh you meant the else clause.

[the remark this answers is not in the document]
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/heroes.dart]: Done. Removed the `mounted` check because it seems redundant with the `attached` check.

**Review by chunhtai — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/widgets/heroes.dart [hunk 8]
```diff
@@ -849,63 +882,89 @@ class HeroController extends NavigatorObserver {
...
+        : _HeroFlightManifest(
+            type: flightType,
+            overlay: overlay,
+            navigatorSize: navigatorRenderObject.size,
+            fromRoute: from,
+            toRoute: to,
+            fromHero: fromHero,
+            toHero: toHero,
+            createRectTween: createRectTween,
+            shuttleBuilder: fromHero.widget.flightShuttleBuilder
```
Comment (HansMuller) [inline packages/flutter/lib/src/widgets/heroes.dart:935]: It looks like this should have been 
```
shuttleBuilder: toHero.widget.flightShuttleBuilder 
  ?? fromHero.widget.flightShuttleBuilder
  ?? _defaultHeroFlightShuttleBuilder,
```
That's the way it was, and the way it was documented. We must be missing a regression test for this case as well.
